File: frontend/src/types/compliance.ts
import type { DocumentUserSummary } from './document';
import type { PaginatedData, SortOrder } from './masterData';

export const complianceStatuses = [
  'COMPLIANT',
  'PARTIALLY_COMPLIANT',
  'NON_COMPLIANT',
  'NEEDS_REVIEW',
  'NOT_EVALUATED',
] as const;

export type ComplianceStatus = (typeof complianceStatuses)[number];

export const complianceJobStatuses = [
  'QUEUED',
  'LOADING_CONTEXT',
  'DETECTING_SECTIONS',
  'GROUPING_CONTENT',
  'VALIDATING_LANGUAGES',
  'VALIDATING_SECTIONS',
  'VALIDATING_ORDER',
  'VALIDATING_TABLES',
  'GENERATING_FINDINGS',
  'CALCULATING_SCORE',
  'PERSISTING',
  'COMPLETED',
  'PARTIALLY_COMPLETED',
  'FAILED',
  'CANCEL_REQUESTED',
  'CANCELLED',
] as const;

export type ComplianceJobStatus = (typeof complianceJobStatuses)[number];

export const terminalComplianceJobStatuses = [
  'COMPLETED',
  'PARTIALLY_COMPLETED',
  'FAILED',
  'CANCELLED',
] as const satisfies readonly ComplianceJobStatus[];

export const activeComplianceJobStatuses = [
  'QUEUED',
  'LOADING_CONTEXT',
  'DETECTING_SECTIONS',
  'GROUPING_CONTENT',
  'VALIDATING_LANGUAGES',
  'VALIDATING_SECTIONS',
  'VALIDATING_ORDER',
  'VALIDATING_TABLES',
  'GENERATING_FINDINGS',
  'CALCULATING_SCORE',
  'PERSISTING',
  'CANCEL_REQUESTED',
] as const satisfies readonly ComplianceJobStatus[];

export type ComplianceJobType =
  'INITIAL_VALIDATION' | 'REVALIDATION' | 'MANUAL_VALIDATION';

export type ComplianceRunStatus = 'COMPLETED' | 'PARTIALLY_COMPLETED' | 'FAILED';
export type RequiredLanguageCode = 'id' | 'en' | 'zh';
export type LanguagePresenceStatus =
  'PRESENT' | 'NOT_PRESENT' | 'INSUFFICIENT_EVIDENCE' | 'MIXED_ONLY';

export interface ComplianceDocumentSummary {
  id: string;
  baseDocumentCode: string;
  title: string;
  departmentId: string;
  departmentName?: string | null;
}

export interface ComplianceRevisionSummary {
  id: string;
  revisionCode: string;
  fullDocumentCode: string;
}

export interface ComplianceFileSummary {
  id: string;
  filename: string;
  fileExtension?: string;
}

export interface ComplianceRuleSummary {
  id: string;
  code: string;
  name: string;
  version?: number;
}

export interface ComplianceStartRequest {
  documentFileId: string;
  extractionRunId?: string | null;
  ocrRunId?: string | null;
  languageDetectionRunId?: string | null;
  validationRuleId?: string | null;
  force?: boolean;
}

export interface ComplianceQueuedResult {
  jobId: string;
  status: ComplianceJobStatus;
  progress: number;
  documentFileId: string;
  runId: string | null;
  reusedExistingResult: boolean;
}

export interface ComplianceJobResultSummary {
  runId: string | null;
  complianceStatus: ComplianceStatus | null;
  complianceScore: number | null;
  totalFindings: number;
  criticalFindings: number;
  majorFindings: number;
  minorFindings: number;
}

export interface ComplianceJob {
  id: string;
  documentId: string;
  documentRevisionId: string;
  documentFileId: string;
  document: ComplianceDocumentSummary | null;
  revision: ComplianceRevisionSummary | null;
  file: ComplianceFileSummary | null;
  validationRule: ComplianceRuleSummary | null;
  jobType: ComplianceJobType;
  status: ComplianceJobStatus;
  progress: number;
  currentStage: string | null;
  requestedBy: DocumentUserSummary | null;
  requestedAt: string;
  startedAt: string | null;
  completedAt: string | null;
  failedAt: string | null;
  cancelledAt: string | null;
  attemptNumber: number;
  maximumAttempts: number;
  errorCode: string | null;
  errorMessage: string | null;
  resultSummary: ComplianceJobResultSummary | null;
}

export interface ComplianceCancelResult {
  id: string;
  status: ComplianceJobStatus;
  progress: number;
  currentStage: string | null;
  cancelledAt: string | null;
}

export interface ComplianceJobListParams {
  search?: string;
  departmentId?: string;
  documentId?: string;
  revisionId?: string;
  documentFileId?: string;
  validationRuleId?: string;
  complianceStatus?: ComplianceStatus;
  status?: ComplianceJobStatus | readonly ComplianceJobStatus[];
  requestedBy?: string;
  requestedFrom?: string;
  requestedTo?: string;
  page: number;
  pageSize: number;
  sortBy?: 'requestedAt' | 'completedAt' | 'status' | 'progress';
  sortOrder?: SortOrder;
}

export type ComplianceJobList = PaginatedData<ComplianceJob>;

export interface ComplianceRun {
  id: string;
  complianceJobId: string;
  documentId: string;
  documentRevisionId: string;
  documentFileId: string;
  extractionRunId: string;
  ocrRunId: string | null;
  languageDetectionRunId: string;
  validationRuleId: string;
  document: ComplianceDocumentSummary | null;
  revision: ComplianceRevisionSummary | null;
  file: ComplianceFileSummary | null;
  validationRule: ComplianceRuleSummary | null;
  ruleSnapshot: Record<string, unknown>;
  sourceContentHash: string;
  status: ComplianceRunStatus;
  complianceStatus: ComplianceStatus;
  complianceScore: number | null;
  maximumScore: number;
  documentCodeScore: number;
  languagePresenceScore: number;
  languageCoverageScore: number;
  sectionCompletenessScore: number;
  languageOrderScore: number;
  translationGroupScore: number;
  tableCompletenessScore: number;
  totalFindings: number;
  criticalFindings: number;
  majorFindings: number;
  minorFindings: number;
  informationFindings: number;
  openFindings: number;
  requiredLanguages: RequiredLanguageCode[];
  detectedLanguages: RequiredLanguageCode[];
  missingLanguages: RequiredLanguageCode[];
  requiredSections: string[];
  detectedSections: string[];
  missingSections: string[];
  warnings: string[];
  metrics: Record<string, unknown>;
  startedAt: string;
  completedAt: string | null;
  requestedBy: DocumentUserSummary | null;
  createdAt: string;
}

export interface LanguageComplianceMetric {
  languageCode: RequiredLanguageCode;
  presence: LanguagePresenceStatus;
  blockCoverage: number;
  characterCoverage: number;
  minimumBlockCoverage: number | null;
  minimumCharacterCoverage: number | null;
  averageConfidence: number | null;
  findingCount: number;
}

export interface TranslationGroupSummary {
  total: number;
  complete: number;
  incomplete: number;
  lowConfidence?: number;
}

export interface ComplianceFindingSummary {
  total: number;
  critical: number;
  major: number;
  minor: number;
  information: number;
  open?: number;
}

export interface ComplianceSummary {
  runId: string;
  status: ComplianceRunStatus;
  complianceStatus: ComplianceStatus;
  complianceScore: number | null;
  requiredLanguages: RequiredLanguageCode[];
  languagePresence: Record<RequiredLanguageCode, LanguagePresenceStatus>;
  languageCoverage: Record<RequiredLanguageCode, number>;
  languageMetrics?: LanguageComplianceMetric[];
  missingLanguages?: RequiredLanguageCode[];
  requiredSections: number;
  detectedSections: number;
  completeSections: number;
  missingSections?: string[];
  translationGroups: TranslationGroupSummary;
  findings: ComplianceFindingSummary;
  warnings: string[];
  prerequisiteErrors?: string[];
}

export interface ScoreComponent {
  earned: number;
  maximum: number;
}

export interface ComplianceScoreBreakdown {
  documentCode: ScoreComponent;
  languagePresence: ScoreComponent;
  languageCoverage: ScoreComponent;
  sectionCompleteness: ScoreComponent;
  languageOrder: ScoreComponent;
  translationGroups: ScoreComponent;
  tableCompleteness: ScoreComponent;
  penalties: {
    major: number;
    minor: number;
    critical?: number;
  };
  scoreCap?: number | null;
  scoreCapReason?: string | null;
  finalScore: number | null;
}

export interface SectionLanguageResult {
  id: string;
  detectedSectionId: string;
  languageCode: RequiredLanguageCode;
  presenceStatus: LanguagePresenceStatus;
  blockCount: number;
  characterCount: number;
  coveragePercentage: number;
  averageConfidence: number | null;
  firstBlockId: string | null;
  lastBlockId: string | null;
  metrics: Record<string, unknown>;
  createdAt: string;
}

export interface DetectedSection {
  id: string;
  complianceRunId: string;
  sectionDefinitionId: string | null;
  canonicalCode: string;
  containerId: string | null;
  startBlockId: string | null;
  endBlockId: string | null;
  headingBlockId: string | null;
  headingText: string;
  headingLanguageCode: RequiredLanguageCode | 'mixed' | 'unknown';
  matchType: string;
  matchConfidence: number;
  sectionOrder: number;
  isRequired: boolean;
  isComplete: boolean;
  languagePresence: Partial<Record<RequiredLanguageCode, LanguagePresenceStatus>>;
  languageResults: SectionLanguageResult[];
  metrics: Record<string, unknown>;
  findingCount: number;
  createdAt: string;
}

export interface ComplianceResultListParams {
  page: number;
  pageSize: number;
}

export type DetectedSectionList = PaginatedData<DetectedSection>;

export type TranslationGroupType =
  | 'HEADING_GROUP'
  | 'PARAGRAPH_GROUP'
  | 'TABLE_ROW_GROUP'
  | 'TABLE_CELL_GROUP'
  | 'XLSX_ROW_GROUP'
  | 'PDF_POSITIONAL_GROUP'
  | 'MANUAL_GROUP';

export interface TranslationGroupMember {
  id: string;
  translationGroupId: string;
  languageCode: RequiredLanguageCode | 'mixed' | 'unknown' | 'other';
  sourceType: string;
  extractedBlockId: string | null;
  ocrBlockId: string | null;
  languageBlockResultId: string | null;
  blockOrder: number;
  textSnapshot: string | null;
  confidence: number;
  position: Record<string, unknown> | null;
  createdAt: string;
}

export interface TranslationGroup {
  id: string;
  complianceRunId: string;
  containerId: string | null;
  detectedSectionId: string | null;
  sectionCode?: string | null;
  groupIndex: number;
  groupType: TranslationGroupType;
  startBlockOrder: number;
  endBlockOrder: number;
  sourceReference: string;
  expectedLanguages: RequiredLanguageCode[];
  detectedLanguages: RequiredLanguageCode[];
  languageOrder: RequiredLanguageCode[];
  actualLanguageOrder?: RequiredLanguageCode[];
  isComplete: boolean;
  isOrderValid: boolean;
  confidence: number;
  metrics: Record<string, unknown>;
  members: TranslationGroupMember[];
  findingCount: number;
  createdAt: string;
}

export interface TranslationGroupListParams extends ComplianceResultListParams {
  isComplete?: boolean;
  isOrderValid?: boolean;
  lowConfidence?: boolean;
  detectedSectionId?: string;
  containerId?: string;
}

export type TranslationGroupList = PaginatedData<TranslationGroup>;

export interface ComplianceHistoryParams {
  page: number;
  pageSize: number;
}

export type ComplianceHistory = PaginatedData<ComplianceRun>;

export interface ComplianceRevalidateRequest {
  reason: string;
  validationRuleId?: string | null;
}

export interface ComplianceComparison {
  currentRunId: string;
  previousRunId: string;
  scoreChange: number;
  previousStatus: ComplianceStatus;
  currentStatus: ComplianceStatus;
  languagesAdded: RequiredLanguageCode[];
  languagesRemoved: RequiredLanguageCode[];
  sectionsAdded: string[];
  sectionsRemoved: string[];
  newFindings: number;
  resolvedCandidates: number;
  repeatedFindings: number;
  translationGroupCompletenessChange: number;
}

export type ComplianceExportFormat = 'json' | 'xlsx';

export interface ComplianceDownload {
  blob: Blob;
  fileName: string | null;
}

export interface ComplianceOverviewParams {
  dateFrom?: string;
  dateTo?: string;
  departmentId?: string;
  sectionId?: string;
  documentTypeId?: string;
  validationRuleId?: string;
  complianceStatus?: ComplianceStatus;
}

export interface ComplianceBreakdownItem {
  label: string;
  total: number;
  compliant: number;
  partiallyCompliant: number;
  nonCompliant: number;
  needsReview: number;
  notEvaluated: number;
}

export interface ComplianceTrendItem {
  period: string;
  score: number;
  validated: number;
}

export interface ComplianceOverview {
  totalValidatedDocuments: number;
  compliant: number;
  partiallyCompliant: number;
  nonCompliant: number;
  needsReview: number;
  notEvaluated: number;
  openCriticalFindings: number;
  openMajorFindings: number;
  byDepartment: ComplianceBreakdownItem[];
  byDocumentType: ComplianceBreakdownItem[];
  trend: ComplianceTrendItem[];
  findingsBySeverity: Record<'critical' | 'major' | 'minor' | 'information', number>;
  missingLanguages: Array<{ languageCode: RequiredLanguageCode; count: number }>;
  missingSections: Array<{ canonicalCode: string; count: number }>;
}

export interface ComplianceReportItem {
  runId: string;
  documentId: string;
  documentFileId: string;
  documentCode: string;
  title: string;
  department: string;
  section: string | null;
  documentType: string;
  revision: string;
  validationRule: string;
  languagePresence: Record<RequiredLanguageCode, LanguagePresenceStatus>;
  sectionCompleteness: number;
  languageOrderValid: boolean | null;
  score: number | null;
  complianceStatus: ComplianceStatus;
  criticalFindings: number;
  majorFindings: number;
  lastValidated: string;
}

export interface ComplianceReportParams extends ComplianceOverviewParams {
  search?: string;
  page: number;
  pageSize: number;
  sortBy?: string;
  sortOrder?: SortOrder;
}

export type ComplianceReport = PaginatedData<ComplianceReportItem>;

export const isTerminalComplianceJobStatus = (status: ComplianceJobStatus): boolean =>
  terminalComplianceJobStatuses.includes(
    status as (typeof terminalComplianceJobStatuses)[number],
  );

export const isActiveComplianceJobStatus = (status: ComplianceJobStatus): boolean =>
  activeComplianceJobStatuses.includes(
    status as (typeof activeComplianceJobStatuses)[number],
  );

File: backend/app/schemas/compliance.py
"""Public Phase 8 compliance job, run, summary, and comparison schemas."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import Field, field_validator

from app.models.compliance_enums import (
    ComplianceJobStatus,
    ComplianceJobType,
    ComplianceRunStatus,
    ComplianceStatus,
)
from app.schemas.base import ApiSchema
from app.schemas.common import PaginationData


def _required_reason(value: str) -> str:
    normalized = value.strip()
    if not normalized:
        raise ValueError("Reason must not be empty.")
    return normalized


class ComplianceStartRequest(ApiSchema):
    document_file_id: UUID
    extraction_run_id: UUID | None = None
    ocr_run_id: UUID | None = None
    language_detection_run_id: UUID | None = None
    validation_rule_id: UUID | None = None
    force: bool = False


class ComplianceQueuedResponse(ApiSchema):
    job_id: UUID
    status: ComplianceJobStatus
    progress: int = Field(ge=0, le=100)
    document_file_id: UUID
    run_id: UUID | None = None
    reused_existing_result: bool = False


class ComplianceRevalidateRequest(ApiSchema):
    reason: str = Field(min_length=1, max_length=2000)
    validation_rule_id: UUID | None = None

    _reason = field_validator("reason", mode="before")(_required_reason)


class ComplianceDocumentReference(ApiSchema):
    id: UUID
    base_document_code: str
    title: str
    department_id: UUID
    department_name: str | None = None


class ComplianceRevisionReference(ApiSchema):
    id: UUID
    revision_code: str
    full_document_code: str


class ComplianceFileReference(ApiSchema):
    id: UUID
    filename: str
    file_extension: str


class ComplianceRuleReference(ApiSchema):
    id: UUID
    code: str
    name: str
    version: int | None = None


class ComplianceRequesterReference(ApiSchema):
    id: UUID
    name: str


class ComplianceJobResponse(ApiSchema):
    id: UUID
    document_id: UUID
    document_revision_id: UUID
    document_file_id: UUID
    extraction_run_id: UUID
    ocr_run_id: UUID | None
    language_detection_run_id: UUID
    validation_rule_id: UUID
    document: ComplianceDocumentReference | None = None
    revision: ComplianceRevisionReference | None = None
    file: ComplianceFileReference | None = None
    validation_rule: ComplianceRuleReference | None = None
    job_type: ComplianceJobType
    status: ComplianceJobStatus
    progress: int = Field(ge=0, le=100)
    current_stage: str | None
    source_content_hash: str | None = None
    requested_by: ComplianceRequesterReference | None
    requested_at: datetime
    started_at: datetime | None
    completed_at: datetime | None
    failed_at: datetime | None
    cancelled_at: datetime | None
    attempt_number: int = Field(ge=1)
    maximum_attempts: int = Field(ge=1)
    error_code: str | None
    error_message: str | None
    error_details: dict[str, Any] | None = None
    result_summary: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime


class ComplianceCancelResponse(ApiSchema):
    id: UUID
    status: ComplianceJobStatus
    progress: int = Field(ge=0, le=100)
    current_stage: str | None = None
    cancelled_at: datetime | None = None


class ComplianceJobListResponse(PaginationData[ComplianceJobResponse]):
    pass


class ComplianceJobFilter(ApiSchema):
    search: str | None = None
    department_id: UUID | None = None
    document_id: UUID | None = None
    document_file_id: UUID | None = None
    validation_rule_id: UUID | None = None
    compliance_status: ComplianceStatus | None = None
    requested_by: UUID | None = None
    status: ComplianceJobStatus | None = None
    requested_from: datetime | None = None
    requested_to: datetime | None = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    sort_by: str = "requestedAt"
    sort_order: str = "desc"


class ComplianceRunResponse(ApiSchema):
    id: UUID
    compliance_job_id: UUID
    document_id: UUID
    document_revision_id: UUID
    document_file_id: UUID
    extraction_run_id: UUID
    ocr_run_id: UUID | None
    language_detection_run_id: UUID
    validation_rule_id: UUID
    document: ComplianceDocumentReference | None = None
    revision: ComplianceRevisionReference | None = None
    file: ComplianceFileReference | None = None
    validation_rule: ComplianceRuleReference | None = None
    rule_snapshot: dict[str, Any]
    source_content_hash: str
    status: ComplianceRunStatus
    compliance_status: ComplianceStatus
    compliance_score: float = Field(ge=0, le=100)
    maximum_score: float = Field(ge=0, le=100)
    document_code_score: float = Field(ge=0)
    language_presence_score: float = Field(ge=0)
    language_coverage_score: float = Field(ge=0)
    section_completeness_score: float = Field(ge=0)
    language_order_score: float = Field(ge=0)
    translation_group_score: float = Field(ge=0)
    table_completeness_score: float = Field(ge=0)
    total_findings: int = Field(ge=0)
    critical_findings: int = Field(ge=0)
    major_findings: int = Field(ge=0)
    minor_findings: int = Field(ge=0)
    information_findings: int = Field(ge=0)
    open_findings: int = Field(ge=0)
    required_languages: list[str]
    detected_languages: list[str] | dict[str, Any]
    missing_languages: list[str]
    required_sections: list[str]
    detected_sections: list[str] | dict[str, Any]
    missing_sections: list[str]
    warnings: list[str] | list[dict[str, Any]]
    metrics: dict[str, Any]
    started_at: datetime
    completed_at: datetime | None
    requested_by: ComplianceRequesterReference | None
    created_at: datetime


class ComplianceRunListResponse(PaginationData[ComplianceRunResponse]):
    pass


class ComplianceLanguageSummary(ApiSchema):
    presence: dict[str, str] = Field(default_factory=dict)
    block_coverage: dict[str, float] = Field(default_factory=dict)
    character_coverage: dict[str, float] = Field(default_factory=dict)
    average_confidence: dict[str, float | None] = Field(default_factory=dict)


class LanguageComplianceMetric(ApiSchema):
    language_code: str
    presence: str
    block_coverage: float = Field(ge=0, le=100)
    character_coverage: float = Field(ge=0, le=100)
    minimum_block_coverage: float | None = Field(default=None, ge=0, le=100)
    minimum_character_coverage: float | None = Field(
        default=None,
        ge=0,
        le=100,
    )
    average_confidence: float | None = Field(default=None, ge=0, le=1)
    finding_count: int = Field(default=0, ge=0)


class ComplianceTranslationGroupSummary(ApiSchema):
    total: int = Field(default=0, ge=0)
    complete: int = Field(default=0, ge=0)
    incomplete: int = Field(default=0, ge=0)
    order_invalid: int = Field(default=0, ge=0)
    low_confidence: int = Field(default=0, ge=0)


class ComplianceFindingSummary(ApiSchema):
    total: int = Field(default=0, ge=0)
    open: int = Field(default=0, ge=0)
    critical: int = Field(default=0, ge=0)
    major: int = Field(default=0, ge=0)
    minor: int = Field(default=0, ge=0)
    information: int = Field(default=0, ge=0)


class ComplianceSummaryResponse(ApiSchema):
    run_id: UUID
    status: ComplianceRunStatus
    compliance_status: ComplianceStatus
    compliance_score: float = Field(ge=0, le=100)
    required_languages: list[str]
    language_presence: dict[str, str] = Field(default_factory=dict)
    language_coverage: dict[str, float] = Field(default_factory=dict)
    language: ComplianceLanguageSummary | None = None
    language_metrics: list[LanguageComplianceMetric] = Field(
        default_factory=list
    )
    required_sections: int = Field(ge=0)
    detected_sections: int = Field(ge=0)
    complete_sections: int = Field(ge=0)
    translation_groups: ComplianceTranslationGroupSummary
    findings: ComplianceFindingSummary
    warnings: list[str] | list[dict[str, Any]] = Field(default_factory=list)


class ComplianceScoreComponent(ApiSchema):
    earned: float = Field(ge=0)
    maximum: float = Field(ge=0)


class ComplianceScorePenalties(ApiSchema):
    major: float = Field(default=0, le=0)
    minor: float = Field(default=0, le=0)
    other: float = Field(default=0, le=0)


class ComplianceScoreBreakdownResponse(ApiSchema):
    document_code: ComplianceScoreComponent
    language_presence: ComplianceScoreComponent
    language_coverage: ComplianceScoreComponent
    section_completeness: ComplianceScoreComponent
    language_order: ComplianceScoreComponent
    translation_groups: ComplianceScoreComponent
    table_completeness: ComplianceScoreComponent
    penalties: ComplianceScorePenalties
    weighted_score: float = Field(ge=0, le=100)
    score_cap: float | None = Field(default=None, ge=0, le=100)
    final_score: float = Field(ge=0, le=100)


class ComplianceComparisonResponse(ApiSchema):
    current_run_id: UUID
    previous_run_id: UUID
    score_change: float
    previous_status: ComplianceStatus
    current_status: ComplianceStatus
    languages_added: list[str] = Field(default_factory=list)
    languages_removed: list[str] = Field(default_factory=list)
    sections_added: list[str] = Field(default_factory=list)
    sections_removed: list[str] = Field(default_factory=list)
    new_findings: int = Field(default=0, ge=0)
    resolved_candidates: int = Field(default=0, ge=0)
    repeated_findings: int = Field(default=0, ge=0)
    translation_group_completeness_change: float | None = None
